From: Irfan Mohammad <irfan@graphiant.com>
Date: Tue, 21 Feb 2023 09:54:56 +0000
Subject: Add SR_SUBSCR_OPER_ONCE option for oper subscr

This option causes a component subscribing to provide operational pull
data to be called no more than once per-subscription for a given get
on the operational datastore.

This can be used to optimise certain usecases. For example, if a
component is subscribed to provide pull state below a list, the default
behaviour causes a component to be called for each element in the list
(unless the request was for a specific element in the list).

As the number of elements in the parent list(s) grows it can take a
significant amount of time, proportional to the time required to actually
collect the data, to issue and handle a callback for each element when
data for all (or a large subset of) elements is requested.

This option therefore gives the capability to handle pull requests in
a "one shot" fashion where it is more performant to do so.
---
 src/modinfo.c          |  18 +++-
 src/sysrepo.c          |   2 +-
 src/sysrepo_types.h    |   9 +-
 tests/test_oper_pull.c | 225 +++++++++++++++++++++++++++++++++++++++++++++++++
 4 files changed, 250 insertions(+), 4 deletions(-)

diff --git a/src/modinfo.c b/src/modinfo.c
index fd656bc..c5374ae 100644
--- a/src/modinfo.c
+++ b/src/modinfo.c
@@ -1158,10 +1158,10 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
             }
 
             /* nested data */
-            for (j = 0; j < set->count; ++j) {
+            if (xpath_subs[0].opts & SR_SUBSCR_OPER_ONCE) {
                 /* get oper data from the client */
                 if ((err_info = sr_xpath_oper_data_get(mod, sub_xpath, request_xpaths, req_xpath_count, orig_name,
-                        orig_data, shm_subs, i, set->dnodes[j], timeout_ms, conn, &oper_data))) {
+                        orig_data, shm_subs, i, NULL, timeout_ms, conn, &oper_data))) {
                     goto cleanup_opergetsub_ext_unlock;
                 }
 
@@ -1170,6 +1170,20 @@ sr_module_oper_data_update(struct sr_mod_info_mod_s *mod, const char *orig_name,
                     lyd_free_all(oper_data);
                     goto cleanup_opergetsub_ext_unlock;
                 }
+            } else {
+                for (j = 0; j < set->count; ++j) {
+                    /* get oper data from the client */
+                    if ((err_info = sr_xpath_oper_data_get(mod, sub_xpath, request_xpaths, req_xpath_count, orig_name,
+                            orig_data, shm_subs, i, set->dnodes[j], timeout_ms, conn, &oper_data))) {
+                        goto cleanup_opergetsub_ext_unlock;
+                    }
+
+                    /* merge into one data tree */
+                    if ((err_info = sr_lyd_merge(data, oper_data, 1, LYD_MERGE_DESTRUCT))) {
+                        lyd_free_all(oper_data);
+                        goto cleanup;
+                    }
+                }
             }
 
 next_iter:
diff --git a/src/sysrepo.c b/src/sysrepo.c
index 2cb639e..ee356b3 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -6963,7 +6963,7 @@ sr_oper_get_subscribe(sr_session_ctx_t *session, const char *module_name, const
 
     conn = session->conn;
     /* only these options are relevant outside this function and will be stored */
-    sub_opts = opts & SR_SUBSCR_OPER_MERGE;
+    sub_opts = opts & (SR_SUBSCR_OPER_MERGE | SR_SUBSCR_OPER_ONCE);
 
     /* CONTEXT LOCK */
     if ((err_info = sr_lycc_lock(conn, SR_LOCK_READ, 0, __func__))) {
diff --git a/src/sysrepo_types.h b/src/sysrepo_types.h
index ceeb496..c08ae33 100644
--- a/src/sysrepo_types.h
+++ b/src/sysrepo_types.h
@@ -451,7 +451,14 @@ typedef enum {
      * event handling but results in 0 filtered-out changes returned by ::sr_module_change_sub_get_info(). Accepted
      * only for ::sr_module_change_subscribe().
      */
-    SR_SUBSCR_FILTER_ORIG = 0x100
+    SR_SUBSCR_FILTER_ORIG = 0x100000,
+
+    /*
+     * @brief For a given "get" operation on the operational datastore, this option ensures that a subscriber
+     * is called no more than once, regardless of whether the subscription path is a child of a list.
+     * When this option is specified, a parent node is not passed to the subscriber's callback.
+     */
+    SR_SUBSCR_OPER_ONCE = 0x100
 
 } sr_subscr_flag_t;
 
diff --git a/tests/test_oper_pull.c b/tests/test_oper_pull.c
index c185275..5c75d7d 100644
--- a/tests/test_oper_pull.c
+++ b/tests/test_oper_pull.c
@@ -3728,6 +3728,230 @@ test_cache_diff(void **state)
 
     assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 6);
 }
+/* TEST */
+static int
+state_once_only_oper_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *xpath,
+        const char *request_xpath, uint32_t request_id, struct lyd_node **parent, void *private_data)
+{
+    struct state *st = (struct state *)private_data;
+    const struct ly_ctx *ly_ctx;
+    struct lyd_node *node;
+
+    (void)request_xpath;
+    (void)request_id;
+    (void)sub_id;
+
+    assert_string_equal(module_name, "mixed-config");
+
+    ly_ctx = sr_acquire_context(sr_session_get_connection(session));
+
+    if (!strcmp(xpath, "/mixed-config:test-state")) {
+        assert_non_null(parent);
+        assert_null(*parent);
+
+        assert_int_equal(LY_SUCCESS, lyd_new_path(NULL, ly_ctx, "/mixed-config:test-state/test-case[name='one']/result", "101", 0, parent));
+        assert_int_equal(LY_SUCCESS, lyd_new_path(*parent, NULL, "/mixed-config:test-state/test-case[name='one']/x", "0.5000", 0, &node));
+        assert_int_equal(LY_SUCCESS, lyd_new_path(*parent, NULL, "/mixed-config:test-state/test-case[name='one']/y", "-0.5000", 0, &node));
+        assert_int_equal(LY_SUCCESS, lyd_new_path(*parent, NULL, "/mixed-config:test-state/test-case[name='one']/z", "-0.2500", 0, &node));
+        assert_int_equal(LY_SUCCESS, lyd_new_path(*parent, NULL, "/mixed-config:test-state/test-case[name='two']", NULL, 0, &node));
+    } else if (parent) {
+        assert_null(*parent);
+        assert_int_equal(LY_SUCCESS, lyd_new_path(NULL, ly_ctx, "/mixed-config:test-state/test-case[name='three']/name", "three", 0, parent));
+        assert_int_equal(LY_SUCCESS, lyd_new_path(*parent, NULL, "/mixed-config:test-state/test-case[name='three']/result", "103", 0, &node));
+        assert_int_equal(LY_SUCCESS, lyd_new_path(*parent, NULL, "/mixed-config:test-state/test-case[name='four']/result", "104", 0, &node));
+        assert_int_equal(LY_SUCCESS, lyd_new_path(*parent, NULL, "/mixed-config:test-state/test-case[name='five']/result", "105", 0, &node));
+    } else {
+        fail();
+    }
+
+    sr_release_context(sr_session_get_connection(session));
+    ATOMIC_INC_RELAXED(st->cb_called);
+    return SR_ERR_OK;
+}
+
+static void
+test_oper_once(void **state)
+{
+    struct state *st = (struct state *)*state;
+    sr_data_t *data;
+    sr_subscription_ctx_t *subscr = NULL;
+    char *str1;
+    const char *str2;
+    int ret;
+
+    /* subscribe as mixed data provider and listen */
+    ret = sr_oper_get_subscribe(st->sess, "mixed-config", "/mixed-config:test-state", state_once_only_oper_cb,
+            st, SR_SUBSCR_OPER_ONCE, &subscr);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* read all state-only data */
+    ret = sr_session_switch_ds(st->sess, SR_DS_OPERATIONAL);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    ATOMIC_STORE_RELAXED(st->cb_called, 0);
+    ret = sr_get_data(st->sess, "/mixed-config:*", 0, 0, SR_OPER_NO_CONFIG | SR_OPER_WITH_ORIGIN, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 1);
+
+    ret = lyd_print_mem(&str1, data->tree, LYD_XML, LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS);
+    assert_int_equal(ret, 0);
+
+    sr_release_data(data);
+
+    str2 =
+    "<test-state xmlns=\"urn:sysrepo:mixed-config\" xmlns:or=\"urn:ietf:params:xml:ns:yang:ietf-origin\" or:origin=\"or:unknown\">"
+        "<test-case>"
+            "<name>one</name>"
+            "<result>101</result>"
+            "<x>0.5</x>"
+            "<y>-0.5</y>"
+            "<z>-0.25</z>"
+        "</test-case>"
+    "</test-state>";
+
+    assert_string_equal(str1, str2);
+    free(str1);
+
+    sr_unsubscribe(subscr);
+    subscr = NULL;
+    /* set some configuration data */
+    ret = sr_session_switch_ds(st->sess, SR_DS_RUNNING);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_set_item_str(st->sess, "/mixed-config:test-state/test-case[name='four']", NULL, NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_set_item_str(st->sess, "/mixed-config:test-state/test-case[name='five']", NULL, NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_set_item_str(st->sess, "/mixed-config:test-state/test-case[name='six']", NULL, NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_apply_changes(st->sess, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* subscribe to all configuration data just to enable them */
+    ret = sr_module_change_subscribe(st->sess, "mixed-config", "/mixed-config:test-state", dummy_change_cb, NULL,
+            0, 0, &subscr);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* subscribe as nested state data provider and listen */
+    ret = sr_oper_get_subscribe(st->sess, "mixed-config", "/mixed-config:test-state/test-case/result", state_once_only_oper_cb,
+            st, SR_SUBSCR_OPER_ONCE, &subscr);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* read some state data */
+    ret = sr_session_switch_ds(st->sess, SR_DS_OPERATIONAL);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    ATOMIC_STORE_RELAXED(st->cb_called, 0);
+    ret = sr_get_data(st->sess, "/mixed-config:test-state/test-case", 0, 0,
+            SR_OPER_NO_CONFIG | SR_OPER_WITH_ORIGIN, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 1);
+
+    ret = lyd_print_mem(&str1, data->tree, LYD_XML, LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS);
+    assert_int_equal(ret, 0);
+
+    sr_release_data(data);
+
+    str2 =
+    "<test-state xmlns=\"urn:sysrepo:mixed-config\" xmlns:or=\"urn:ietf:params:xml:ns:yang:ietf-origin\" or:origin=\"or:intended\">"
+        "<test-case>"
+            "<name>five</name>"
+            "<result>105</result>"
+        "</test-case>"
+        "<test-case>"
+            "<name>four</name>"
+            "<result>104</result>"
+        "</test-case>"
+        "<test-case>"
+            "<name>three</name>"
+            "<result>103</result>"
+        "</test-case>"
+    "</test-state>";
+
+    assert_string_equal(str1, str2);
+    free(str1);
+
+    /* read some state and configuration data */
+    ATOMIC_STORE_RELAXED(st->cb_called, 0);
+    ret = sr_get_data(st->sess, "/mixed-config:test-state/test-case", 0, 0,
+            SR_OPER_WITH_ORIGIN, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 1);
+
+    ret = lyd_print_mem(&str1, data->tree, LYD_XML, LYD_PRINT_SHRINK| LYD_PRINT_WITHSIBLINGS);
+    assert_int_equal(ret, 0);
+
+    sr_release_data(data);
+
+    str2 =
+    "<test-state xmlns=\"urn:sysrepo:mixed-config\" xmlns:or=\"urn:ietf:params:xml:ns:yang:ietf-origin\" or:origin=\"or:intended\">"
+        "<test-case>"
+            "<name>five</name>"
+            "<result>105</result>"
+        "</test-case>"
+        "<test-case>"
+            "<name>four</name>"
+            "<result>104</result>"
+        "</test-case>"
+        "<test-case>"
+            "<name>six</name>"
+        "</test-case>"
+        "<test-case>"
+            "<name>three</name>"
+            "<result>103</result>"
+        "</test-case>"
+    "</test-state>";
+
+    assert_string_equal(str1, str2);
+    free(str1);
+
+    /* read a subset of data which is in both running and operational stores */
+    ATOMIC_STORE_RELAXED(st->cb_called, 0);
+    ret = sr_get_data(st->sess, "/mixed-config:test-state/test-case[name='four']", 0, 0,
+            SR_OPER_NO_CONFIG | SR_OPER_WITH_ORIGIN, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 1);
+
+    ret = lyd_print_mem(&str1, data->tree, LYD_XML, LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS);
+    assert_int_equal(ret, 0);
+
+    sr_release_data(data);
+
+    str2 =
+    "<test-state xmlns=\"urn:sysrepo:mixed-config\" xmlns:or=\"urn:ietf:params:xml:ns:yang:ietf-origin\" or:origin=\"or:intended\">"
+        "<test-case>"
+            "<name>four</name>"
+            "<result>104</result>"
+        "</test-case>"
+    "</test-state>";
+
+    assert_string_equal(str1, str2);
+    free(str1);
+
+    /* read a subset of state-only data */
+    ATOMIC_STORE_RELAXED(st->cb_called, 0);
+    ret = sr_get_data(st->sess, "/mixed-config:test-state/test-case[name='three']", 0, 0,
+            SR_OPER_NO_CONFIG | SR_OPER_WITH_ORIGIN, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 1);
+
+    ret = lyd_print_mem(&str1, data->tree, LYD_XML, LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS);
+    assert_int_equal(ret, 0);
+
+    sr_release_data(data);
+
+    str2 =
+    "<test-state xmlns=\"urn:sysrepo:mixed-config\" xmlns:or=\"urn:ietf:params:xml:ns:yang:ietf-origin\" or:origin=\"or:intended\">"
+        "<test-case>"
+            "<name>three</name>"
+            "<result>103</result>"
+        "</test-case>"
+    "</test-state>";
+
+    assert_string_equal(str1, str2);
+    free(str1);
+
+    sr_unsubscribe(subscr);
+}
 
 int
 main(void)
@@ -3759,6 +3983,7 @@ main(void)
         cmocka_unit_test_teardown(test_cache, clear_up),
         cmocka_unit_test_teardown(test_cache_no_sub, clear_up),
         cmocka_unit_test_teardown(test_cache_diff, clear_up),
+        cmocka_unit_test_teardown(test_oper_once, clear_up),
     };
 
     setenv("CMOCKA_TEST_ABORT", "1", 1);
